Ensure WP_Text_Diff_Renderer_Table is PHP 8.2 compatible

anton-vlasenko opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

This PR aims to add missing class properties to ensure `WP_Text_Diff_Renderer_Table` is PHP 8.2 compatible.

Trac ticket: https://core.trac.wordpress.org/ticket/59431

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

anton-vlasenko committed:
Add new PHP properties on the WP_Text_Diff_Renderer_Table class to ensure compatbility with PHP 8.2.
PHP 8.2 doesn't support dynamic properties.

anton-vlasenko commented on src/wp-includes/class-wp-text-diff-renderer-table.php:
```diff
@@ -30,6 +30,30 @@ class WP_Text_Diff_Renderer_Table extends Text_Diff_Renderer {
+	 * @since 6.4.0
+	 * @var string|null
+	 */
+	public $_title;
```
I intentionally didn't initialize these class properties because they used to be set to `NULL` in the previous implementation. This is to maintain BC.

hellofromtonya replied:
Yes, I agree, the original default value of `null` is being maintained.

anton-vlasenko committed:
Add description.

mukeshpanchal27 commented:
Thank you, @anton-vlasenko, for the PR. Could you also include a brief description indicating that we added these variables for backward compatibility (BC)? This way, we can prevent someone from raising a ticket in the future to remove these variables, claiming they are unused.

anton-vlasenko committed:
Add more detailed description to class properties.

anton-vlasenko commented:
> Thank you, @anton-vlasenko, for the PR. Could you also include a brief description indicating that we added these variables for backward compatibility (BC)?

Thank you for checking this PR, @mukeshpanchal27!
I've added more detailed descriptions to the class properties in 73ebb738b7c1c30485472604051627cdc5d7a9b6 to clarify their purpose.

>This way, we can prevent someone from raising a ticket in the future to remove these variables, claiming they are unused.

If I understand you correctly, you're referring to this [comment](https://github.com/WordPress/wordpress-develop/pull/5453#discussion_r1352881551). These class properties weren't just declared to maintain BC. They were introduced to ensure compatibility with PHP 8.2. What I meant is that these properties are not assigned any value to ensure BC. In my opinion, this doesn't need to be explained in the code. But that's my opinion.

hellofromtonya commented on src/wp-includes/class-wp-text-diff-renderer-table.php:
```diff
@@ -30,6 +30,30 @@ class WP_Text_Diff_Renderer_Table extends Text_Diff_Renderer {
+	/**
+	 * Title of the item being compared.
+	 *
+	 * @since 6.4.0 Explicitly declare dynamic class property to ensure PHP 8.2 compatibility.
```
The `@since` explanation is not required, though it can help to share why it was added.

What about the explanation itself? 
I think it can safely be shortened to remove the PHP part. Why? With the changes made to magic methods, all non-declared properties raise a deprecation (when in `WP_DEBUG` is on). It's not limited to PHP 8.2 and newer. Thus, explaining it is for PHP 8.2 compatibility will likely cause confusion in the future.

```suggestion
	 * @since 6.4.0 Declared a previously dynamic property.
```

hellofromtonya commented on src/wp-includes/class-wp-text-diff-renderer-table.php:
```diff
@@ -30,6 +30,30 @@ class WP_Text_Diff_Renderer_Table extends Text_Diff_Renderer {
+	/**
+	 * Title for the right column.
+	 *
+	 * @since 6.4.0 Explicitly declare dynamic class property to ensure PHP 8.2 compatibility.
```
```suggestion
	 * @since 6.4.0 Declared a previously dynamic property.
```

hellofromtonya commented on src/wp-includes/class-wp-text-diff-renderer-table.php:
```diff
@@ -30,6 +30,30 @@ class WP_Text_Diff_Renderer_Table extends Text_Diff_Renderer {
+	/**
+	 * Title for the left column.
+	 *
+	 * @since 6.4.0 Explicitly declare dynamic class property to ensure PHP 8.2 compatibility.
```
```suggestion
	 * @since 6.4.0 Declared a previously dynamic property.
```

hellofromtonya committed:
Shorten since explanation.

hellofromtonya reviewed (approved):
Core is using these 3 properties, which prior to this patch, were dynamic properties. With the changes to the magic methods, dynamic properties are deprecated for all PHP versions. Explicitly declaring these properties on the class is a follow-up to that scope of work and is needed.

Changes LTGM 👍  Ready for commit.

hellofromtonya commented:
To follow-up on the discussion of these properties:

Anton is right - they are not being declared on the class for BC. Rather,  this patch is necessary to explicitly declare the properties that are being used in Core.

Why? Prior to this patch, the properties were dynamic and thus were accessible only through the magic methods. [r56354](https://core.trac.wordpress.org/changeset/56354) modified the magic methods for all WP supported PHP versions (not just PHP 8.2+) to raise a deprecation for all dynamic properties.

Core itself though should not be raising this new deprecation as it's properties should all be declared.

hellofromtonya commented:
I first reviewed this patch thinking - yes, let's add the properties as declared. But then looking at [r56354](https://core.trac.wordpress.org/changeset/56354) I remember the changes implemented uses the `$compat_fields` property to access these "declared" properties (compatible fields) through the magic methods.

I wondering .. should these 3 properties also be added to the `$compat_fields` array of compatible fields, rather than declaring them on the class? In this way, the current behavior is maintained, i.e. these 3 will still use the magic methods as they always have.

@anton-vlasenko @mukeshpanchal27  what do you think?

anton-vlasenko commented:
@hellofromtonya I'm looking at https://core.trac.wordpress.org/ticket/56034 and, as far as I understand, the recommended way of dealing with known, named dynamic properties is to declare them on the class. 
I'm referring to this part of the ticket (highlighted text): 
![Screenshot 2023-10-11 at 19 08 30](https://github.com/WordPress/wordpress-develop/assets/43744263/bd16bdf4-20b6-403a-a4ce-cd6798ba94aa) 
However, I agree that there is probably less risk in adding these properties to the `$compat_fields` array than in declaring them on the class. I need to investigate potential BC risks.

hellofromtonya commented:
@anton-vlasenko you're right. That is the recommended approach. Thanks for reminding me!

I'm curious of the previous behavior, i.e. before [r56354](https://core.trac.wordpress.org/changeset/56354). Let's double check before moving forward with committing this change.

anton-vlasenko commented:
Thank you, @hellofromtonya.
Before https://core.trac.wordpress.org/changeset/56354, these properties were dynamic and were never set in the `Text_Diff_Renderer`'s constructor because `__isset()` returned false. Now, the behavior remains the same. The only difference is that these properties are now declared, so the `__isset()` magic method is no longer called for them.
IMO, having them declared on the class is fine. But I'm open to hearing alternative opinions.

hellofromtonya commented:
Thanks @anton-vlasenko. Testing before and after for these Core specific dynamic properties:

* Before 6.4: they always returned `null` and could not be changed / set https://3v4l.org/FrRJL
* But this PR changes that behavior by allowing them to now be set, but ... that is the [plan per proposal ](https://core.trac.wordpress.org/ticket/56034)on how to deal with "known named dynamic properties".

>Known, named, dynamic property | Declare the property on the (parent) class

So the PR as it is now aligns to that plan.

hellofromtonya commented:
Committed via https://core.trac.wordpress.org/changeset/56938.
